node/pkg/cni/token_watch.go

```diff
@@ -1,56 +1,158 @@
+		Spec: authenticationv1.TokenRequestSpec{
+			Audiences:         []string{"kubernetes"},
+			ExpirationSeconds: &validity,
+		},
+	}
+
+	tokenRequest, err := clientset.CoreV1().ServiceAccounts(t.namespace).CreateToken(context.TODO(), t.serviceAccountName, tr, metav1.CreateOptions{})
+	if apierrors.IsNotFound(err) && !t.tokenRequestSupported(clientset) {
+		logrus.WithError(err).Error("Unable to create token for CNI kubeconfig as token request api is not supported")
+		return tokenUpdateError(errors.New("token request api not supported"))
+	}
 	if err != nil {
-		panic(err)
+		logrus.WithError(err).Error("Unable to create token for CNI kubeconfig")
+		return tokenUpdateError(err)
 	}
-	defer watcher.Close()
 
-	// Watch for changes to the serviceaccount directory. Rety if necessary.
+	return TokenUpdate{
+		Token:          tokenRequest.Status.Token,
+		ExpirationTime: tokenRequest.Status.ExpirationTimestamp.Time,
+	}
+}
+
+func (t *TokenRefresher) TokenChan() <-chan TokenUpdate {
+	return t.tokenChan
+}
+
+func (t *TokenRefresher) Run() {
+	var nextExpiration time.Time
 	for {
-		if err := watcher.Add(serviceaccountDirectory); err != nil {
-			// Error watching the file - retry
-			logrus.WithError(err).Error("Failed to watch Kubernetes serviceaccount files.")
-			time.Sleep(5 * time.Second)
-			continue
+		tu := t.UpdateToken()
+		if tu.Error == nil {
+			nextExpiration = tu.ExpirationTime
+			t.tokenChan <- tu
+		}
+		now := time.Now()
+		var sleepTime time.Duration
+		// Do some basic rate limiting to prevent flooding the kube apiserver with requests
+		if nextExpiration.Before(now.Add(minTokenRetryDuration * defaultRefreshFraction)) {
+			sleepTime = minTokenRetryDuration
+		} else {
+			sleepTime = nextExpiration.Sub(now) / defaultRefreshFraction
 		}
+		if logrus.IsLevelEnabled(logrus.DebugLevel) {
+			logrus.Debugf("Going to sleep for %s", sleepTime.String())
+		}
+		time.Sleep(sleepTime)
+	}
+}
 
-		// Successfully watched the file. Break from the retry loop.
-		logrus.WithField("directory", serviceaccountDirectory).Info("Watching contents for changes.")
-		break
+func (t *TokenRefresher) tokenRequestSupported(clientset *kubernetes.Clientset) bool {
+	t.tokenOnce.Do(func() {
+		resources, err := clientset.Discovery().ServerResourcesForGroupVersion("v1")
+		if err != nil {
+			return
+		}
+		for _, resource := range resources.APIResources {
+			if resource.Name == "serviceaccounts/token" {
+				t.tokenSupported = true
+				return
+			}
+		}
+	})
+	return t.tokenSupported
+}
+
+func tokenUpdateError(err error) TokenUpdate {
+	return TokenUpdate{
+		Error: err,
 	}
+}
 
-	// Handle events from the watcher.
-	for {
-		// To prevent tight looping, add a sleep here.
-		time.Sleep(1 * time.Second)
+func Run() {
+	// Log to stdout.  this prevents our logs from being interpreted as errors by, for example,
+	// fluentd's default configuration.
+	logrus.SetOutput(os.Stdout)
+
+	tr := NewTokenRefresher(metav1.NamespaceSystem, "calico-node")
```

Comment:
We should use the namespace file mounted into the pod's filesystem to determine the namespace we're running in - not all calico/node pods run in kube-system (e.g., operator-based deployments use calico-system).

Reply:
Here's an example: https://github.com/tigera/operator/blob/ac07c89ead29e2adf3907b00b1b07a53088e8720/pkg/common/operator_namespace.go#L32-L38

Reply:
Implemented the file based namespace lookup as suggested.